src/Nethermind/Nethermind.Optimism/Rpc/OptimismTransactionForRpc.cs

```diff
@@ -3,27 +3,91 @@
+/// </remarks>
+public class OptimismTransactionForRpc : TransactionForRpc, IFromTransaction<OptimismTransactionForRpc>
 {
-    public OptimismTransactionForRpc(Hash256? blockHash, OptimismTxReceipt? receipt, Transaction transaction, UInt256? baseFee = null)
-       : base(blockHash, receipt?.BlockNumber, receipt?.Index, transaction, baseFee)
-    {
-        if (transaction.Type == TxType.DepositTx)
-        {
-            SourceHash = transaction.SourceHash;
-            Mint = transaction.Mint;
-            IsSystemTx = transaction.IsOPSystemTransaction ? true : null;
-            Nonce = receipt?.DepositNonce;
-            DepositReceiptVersion = receipt?.DepositReceiptVersion;
-        }
-    }
+    public static TxType TxType => TxType.DepositTx;
+
+    public override TxType? Type => TxType;
+
+    public Hash256? SourceHash { get; set; }
+
+    public Address? From { get; set; }
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public Address? To { get; set; }
+
+    public UInt256? Mint { get; set; }
+
+    public UInt256? Value { get; set; }
+
+    public ulong? Gas { get; set; }
+
+    public bool? IsSystemTx { get; set; }
+
+    public byte[]? Input { get; set; }
+
+    public UInt256? Nonce { get; set; }
 
+    #region Nethermind specific fields
     [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
     public UInt256? DepositReceiptVersion { get; set; }
+    #endregion
+
+    [JsonConstructor]
+    public OptimismTransactionForRpc() { }
+
+    public OptimismTransactionForRpc(Transaction transaction, int? txIndex = null, Hash256? blockHash = null, long? blockNumber = null, OptimismTxReceipt? receipt = null)
+        : base(transaction, txIndex, blockHash, blockNumber)
+    {
+        SourceHash = transaction.SourceHash ?? Hash256.Zero;
+        From = transaction.SenderAddress ?? Address.SystemUser;
+        To = transaction.To;
+        Mint = transaction.Mint;
+        Value = transaction.Value;
+        // TODO: Unsafe cast
+        Gas = (ulong)transaction.GasLimit;
+        IsSystemTx = transaction.IsOPSystemTransaction;
+        Input = transaction.Data?.ToArray() ?? [];
+        Nonce = receipt?.DepositNonce ?? 0;
+
+        DepositReceiptVersion = receipt?.DepositReceiptVersion;
+    }
+
+    public override Transaction ToTransaction()
+    {
+        var tx = base.ToTransaction();
+
+        tx.SourceHash = SourceHash ?? throw new ArgumentNullException(nameof(SourceHash));
+        tx.SenderAddress = From ?? throw new ArgumentNullException(nameof(From));
+        tx.To = To;
+        tx.Mint = Mint ?? 0;
+        tx.Value = Value ?? throw new ArgumentNullException(nameof(Value));
+        // TODO: Unsafe cast
+        tx.GasLimit = (long)(Gas ?? throw new ArgumentNullException(nameof(Gas)));
+        tx.IsOPSystemTransaction = IsSystemTx ?? false;
+        tx.Data = Input ?? throw new ArgumentNullException(nameof(Input));
```

Comment:
According to the official OP repo these fields are required and we should reject transactions that don't provide it:

https://github.com/ethereum-optimism/op-geth/blob/c283254e5447f127e3b9350860985911dab9cd2f/core/types/transaction_marshalling_test.go#L27-L67

Reply:
Deposit transaction should never come from RPC? Maybe we should just fail here.